pkgs/development/haskell-modules/setup.nix

```diff
@@ -0,0 +1,42 @@
+# This is the generic `Setup` installer for haskell packages.
+#
+# For now we just build the `Setup.hs` against the 
+{ stdenv, ghc, lib }:
+{ setupHaskellDepends ? []
+, setupHs ? ''
+  import Distribution.Simple
+  main = defaultMain
+''
+} @ args:
+stdenv.mkDerivation (rec {
+  name = "Setup";
+
+  phases = [ "buildPhase" ];
+
+  buildInputs = [ ghc ] ++ setupHaskellDepends;
+  LANG = "en_US.UTF-8";
+
+  # build up the package database, with the additional dependencies 
+  preBuild = ''
+    setupPackageConfDir="$TMPDIR/setup-package.conf.d"
+    mkdir -p "$setupPackageConfDir"
+
+    # Add the setupHaskellDepends to a custom package-db,
+    # which ghc will be provided with.
+
+    for p in ${lib.escapeShellArgs setupHaskellDepends}; do
```

Comment:
Here propagated setup dependencies are not taken into account. That is, your pkgdb will contain only immediate setup dependencies but their transitive dependencies will not be in it, so it won’t compile.